BacktraceScreen.c

```diff
@@ -0,0 +1,419 @@
+         printingHelper->maxDemangledFuncNameLen > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFuncNameLen;
+      if (frame->demangleFunctionName) {
+         functionName = frame->demangleFunctionName;
+      }
+   }
+
+   char *completeFunctionName = NULL;
+   xAsprintf(&completeFunctionName, "%s+0x%zx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjNameLen;
+   if (!!(displayOptions & SHOW_FULL_PATH_OBJECT)) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjPathLen;
+   }
+
+   size_t maxAddrLen = printingHelper->maxAddrLen - strlen("0x");
+   char* line = NULL;
+   int objectPathStart = -1;
+
+   /*
+    * The parameters for printf are of type int.
+    * A check is needed to prevent integer overflow.
+    */
+   assert(printingHelper->maxFrameNumLen <= INT_MAX);
+   assert(maxAddrLen <= INT_MAX);
+   assert(maxFunctionNameLength <= INT_MAX);
+   assert(objectLength <= INT_MAX);
+   int len = xAsprintf(&line, "%*d 0x%0*zx %-*s %n%-*s",
+      (int)printingHelper->maxFrameNumLen, frame->index,
+      (int)maxAddrLen, frame->address,
+      (int)maxFunctionNameLength, completeFunctionName,
+      &objectPathStart,
+      (int)objectLength, objectDisplayed ? objectDisplayed : "-"
+   );
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   free(completeFunctionName);
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
+   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
+}
+
+void BacktraceFrame_delete(Object* object) {
+   BacktraceFrame* this = (BacktraceFrame*)object;
+   free(this->functionName);
+   free(this->demangleFunctionName);
+   free(this->objectPath);
+   free(this->objectName);
+   free(this);
+}
+
+static void BacktracePanel_setError(BacktracePanel* this, const char* error) {
+   assert(error);
+   assert(!this->error);
+
+   Panel* super = &this->super;
+   Panel_prune(super);
+
+   Vector* lines = this->super.items;
+   if (lines) {
+      Vector_delete(lines);
+   }
+
+   this->super.items = Vector_new(Class(ListItem), true, DEFAULT_SIZE);
+   Panel_set(super, 0, (Object*)ListItem_new(error, 0));
+}
+
+static void BacktracePanel_makePrintingHelper(const BacktracePanel* this, BacktracePanelPrintingHelper* printingHelper) {
+   Vector* lines = this->super.items;
+   size_t longestAddress = 0;
+
+   for (int i = 0; i < Vector_size(lines); i++) {
+      BacktraceFrame* frame = (BacktraceFrame*)Vector_get(lines, i);
+      frame->backtracePanel = this;
```

Comment:
What does this line do in particular? An issue I just found out is this break the `const` assumption of the `BacktracePanel* this`.

Reply:
The BacktracePanel contains all BacktraceFrame instances, making frames easily accessible via the items attribute. However, accessing the panel from a frame is more complex. To address this, I added a weak reference from each frame to the panel, allowing panel access (e.g., for settings) when only a frame object is available. It is const because it ensures the panel isn’t modified or freed (without a cast).

Reply:
I'm not sure where you need the BacktracePanel reference from the BacktraceFrame, but the `makePrintingHelper` function is supposed to do one thing - the preparation of the "printing helper" data. It's not supposed to modify the BacktraceFrame objects.

If you do need the references from the frame, I suggest you set the references when the frame objects are allocated. This might require you to modify the `BacktraceFrame_new()` function to take one parameter, for example.

Reply:
Yes It's true, it's a good observation. Introducing `void BacktraceFrame_new(Panel *panel)` says clearly that is a weak dependency, but it also means that I need to pass a `Panel *` to `Platform_getBacktrace`.